utils/v2/connector/impl/RedStonePriceFeed.ts

```diff
@@ -1,47 +1,116 @@
-      + (asset ? `&symbol=${asset}` : '');
-    
+      // Getting default data sources config for provider if not specified
+      if (!this.priceFeedOptions.dataSources) {
+        this.priceFeedOptions.dataSources = getDefaultDataSourceConfig(providerId);
+      }
+
+      // Init fetchers
+      for (const [i, source] of this.priceFeedOptions.dataSources!.sources!.entries()) {
+        if (priceFeedOptions.asset && source.disabledForSinglePrices) {
+          console.log(`Skipping ${i} (${source.type}) source init`);
+        } else {
+          const fetcherForSource = createFetcher(source, priceFeedOptions.asset);
+          this.fetchers.push(fetcherForSource);
+        }
+      }
+
+      // TODO: get rid of it, it's a potential single point of failure
+      this.getSigner(); // we are loading signer public key in advance
   }
 
+  // This is the entrypoint function of this module
   async getSignedPrice(): Promise<SignedPriceDataType> {
-    const response = await axios.get(this.apiUrl);
+    // We need to get signer public key firstly
+    // It will be used to pre-verify signatures off-chain
+    await this.getSigner();
+    
+    // Fetching data from all sources simultaneously with timeout
+    const timeoutMilliseconds =
+      this.priceFeedOptions.dataSources?.timeoutMilliseconds
+      || DEFAULT_TIMEOUT_MILLISECONDS;
+    const promises = this.fetchers.map(fetcher =>
+      fetcher.getLatestDataWithTimeout(this.providerId, timeoutMilliseconds));
+    const results = await Promise.allSettled(promises);
+
+    // Validating fetched data
+    const fulfilledPromisesResults = results.filter(r => r.status === "fulfilled") as
+      PromiseFulfilledResult<SignedDataPackageResponse>[];
+    const dataPackages = fulfilledPromisesResults.map(r => r.value);
+    const validDataPackages = dataPackages.filter(p =>
+      validateDataPackage(
+        p,
+        this.priceFeedOptions,
+        this.cachedSigner!,
+      )
+    );
 
-    const pricePackage = _.pick(response.data, ["prices", "timestamp"]);
-    const serialized = this.priceSigner.serializeToMessage(pricePackage);
+    if (validDataPackages.length === 0) {
+      console.error(results);
+      throw new Error(`Failed to load valid data packages`);
+    }
 
-    // TODO: change return type of the priceSigner.serializeToMessage?
-    // in the end, we're using TYPEScript here ;-)
-    const priceData: PriceDataType = serialized as PriceDataType;
+    // Selecting the final value
+    const finalResponse = selectDataPackage(
+      validDataPackages,
+      this.priceFeedOptions.dataSources!.valueSelectionAlgorithm);
 
-    return {
-      priceData,
-      ..._.pick(response.data, ["signer", "signature", "liteSignature"]),
-    };
+    return convertResponseToPricePackage(finalResponse, this.cachedSigner!);
   }
 
+  // TODO: get rid of it later
+  // It is a potential single point of failure
```

Comment:
But the signer is only needed for a pre-validation. 
We could always skip it and rely only on the on-chain validation

Reply:
That's true, but if we use many sources, it would make sense to validate it before sending potentially invalid transaction. Of course it depends on the time costs of offchain signature verification. So I'll do the performance tests